core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -546,24 +546,15 @@ assert getByAllocationId(failedShard.shardId(), failedShard.allocationId().getId
             assert failedShard.active();
             if (failedShard.primary()) {
                 // promote active replica to primary if active replica exists
-                ShardRouting candidate = activeReplica(failedShard.shardId());
-                if (candidate == null) {
+                ShardRouting activeReplica = activeReplica(failedShard.shardId());
+                if (activeReplica == null) {
                     moveToUnassigned(failedShard, unassignedInfo);
                 } else {
+                    // if the activeReplica was relocating before this call to failShard, its relocation was cancelled above when we
+                    // failed initializing replica shards (and moved replica relocation source back to started)
+                    assert activeReplica.started() : "replica relocation should have been cancelled: " + activeReplica;
                     movePrimaryToUnassignedAndDemoteToReplica(failedShard, unassignedInfo);
-                    ShardRouting primarySwappedCandidate = promoteAssignedReplicaShardToPrimary(candidate);
-                    if (primarySwappedCandidate.relocatingNodeId() != null) {
-                        // its also relocating, make sure to move the other routing to primary
-                        RoutingNode node = node(primarySwappedCandidate.relocatingNodeId());
-                        if (node != null) {
-                            for (ShardRouting shardRouting : node) {
-                                if (shardRouting.shardId().equals(primarySwappedCandidate.shardId()) && !shardRouting.primary()) {
-                                    promoteAssignedReplicaShardToPrimary(shardRouting);
-                                    break;
-                                }
-                            }
-                        }
-                    }
+                    ShardRouting primarySwappedCandidate = promoteActiveReplicaShardToPrimary(activeReplica);
```

Comment:
I like the variable name changes, makes it clearer to understand the purpose of the object.